servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/DefaultAsyncContextProvider.java

```diff
@@ -104,165 +104,115 @@ public final CompletableSource.Subscriber wrapCancellable(final CompletableSourc
+        return new ContextPreservingCompletableSubscriber(subscriber, context, null);
     }
 
     @Override
     public final CompletableSource.Subscriber wrapCompletableSubscriber(final CompletableSource.Subscriber subscriber,
                                                                         final CapturedContext context) {
-        if (subscriber instanceof ContextPreservingCancellableCompletableSubscriber) {
-            final ContextPreservingCancellableCompletableSubscriber s =
-                    (ContextPreservingCancellableCompletableSubscriber) subscriber;
-            if (s.capturedContext == context) {
-                // replace current wrapper with wrapper that includes Subscriber and Cancellable
-                return new ContextPreservingCompletableSubscriberAndCancellable(s.subscriber, context);
-            }
-        } else if (subscriber instanceof ContextPreservingCompletableSubscriber &&
-                ((ContextPreservingCompletableSubscriber) subscriber).capturedContext == context) {
-            // no need to check for instanceof ContextPreservingCompletableSubscriberAndCancellable, because
-            // it extends from ContextPreservingCompletableSubscriber.
-            return subscriber;
+        if (subscriber instanceof ContextPreservingCompletableSubscriber) {
+            final ContextPreservingCompletableSubscriber s = (ContextPreservingCompletableSubscriber) subscriber;
+            return s.subscriberCapturedContext != null ? subscriber :
+                    new ContextPreservingCompletableSubscriber(s.subscriber, s.cancellableCapturedContext, context);
         }
-        return new ContextPreservingCompletableSubscriber(subscriber, context);
+        return new ContextPreservingCompletableSubscriber(subscriber, null, context);
     }
 
     @Override
     public final CompletableSource.Subscriber wrapCompletableSubscriberAndCancellable(
             final CompletableSource.Subscriber subscriber, final CapturedContext context) {
         if (subscriber instanceof ContextPreservingCompletableSubscriber) {
             final ContextPreservingCompletableSubscriber s = (ContextPreservingCompletableSubscriber) subscriber;
-            if (s.capturedContext == context) {
-                return subscriber instanceof ContextPreservingCompletableSubscriberAndCancellable ? subscriber :
-                        new ContextPreservingCompletableSubscriberAndCancellable(s.subscriber, context);
-            }
-        } else if (subscriber instanceof ContextPreservingCancellableCompletableSubscriber) {
-            final ContextPreservingCancellableCompletableSubscriber s =
-                    (ContextPreservingCancellableCompletableSubscriber) subscriber;
-            if (s.capturedContext == context) {
-                return new ContextPreservingCompletableSubscriberAndCancellable(s.subscriber, context);
+            if (s.cancellableCapturedContext == context && s.subscriberCapturedContext != null) {
```

Comment:
Consider assing a similar comment as you have in `wrapCancellable` to clarify why equality check here is relevant